asymmetric matcher type definitions are not good (yet again)

After the recent change, cannot be used as nested in `jest`.
See [this comment](https://github.com/jest-community/jest-extended/commit/4b3fa65011cd622bfc44b29d646cc8b34d4c4a93#r95426283).

@rluvaton negative. now I have
```
  ● Test suite failed to run

    test.spec.ts:17:7 - error TS2322: Type 'AsymmetricMatcher' is not assignable to type 'number'.

    17       baz: expect.toBeNumber()
             ~~~

      test.spec.ts:5:3
        5   baz?: number;
            ~~~
        The expected type comes from property 'baz' which is declared here on type 'Foo'

```

--- a/examples/typescript/all/test/add.test.ts
+++ b/examples/typescript/all/test/add.test.ts
@@ -25,4 +25,14 @@ describe('add', () => {
   test('adding two odd numbers gives an even number', () => {
     expect(add(1, 3)).toBeEven();
   });
+
+  test('array equal another array regardless of the order', () => {
+    expect([1, 3]).toIncludeSameMembers([3, 1]);
+  });
+
+  test('array property equal another array regardless of the order', () => {
+    expect({ array: [1, 3] }).toEqual({
+      array: expect.toIncludeSameMembers([3, 1]),
+    });
+  });
 });

--- a/types/index.d.ts
+++ b/types/index.d.ts
@@ -1,13 +1,13 @@
 /* eslint-disable @typescript-eslint/no-explicit-any */
 
-interface CustomMatchers {
+interface CustomMatchers<R> {
   /**
    * Note: Currently unimplemented
    * Passing assertion
    *
    * @param {String} message
    */
-  pass(message: string): void;
+  pass(message: string): R;
 
   /**
    * Note: Currently unimplemented
@@ -15,128 +15,128 @@ interface CustomMatchers {
    *
    * @param {String} message
    */
-  fail(message: string): void;
+  fail(message: string): R;
 
   /**
    * Use .toBeEmpty when checking if a String '', Array [] or Object {} is empty.
    */
-  toBeEmpty(): void;
+  toBeEmpty(): R;
 
   /**
    * Use .toBeOneOf when checking if a value is a member of a given Array.
    * @param {Array.<*>} members
    */
-  toBeOneOf<E = unknown>(members: readonly E[]): void;
+  toBeOneOf<E = unknown>(members: readonly E[]): R;
 
   /**
    * Use `.toBeNil` when checking a value is `null` or `undefined`.
    */
-  toBeNil(): void;
+  toBeNil(): R;
 
   /**
    * Use `.toSatisfy` when you want to use a custom matcher by supplying a predicate function that returns a `Boolean`.
    * @param {Function} predicate
    */
-  toSatisfy<E = any>(predicate: (x: E) => boolean): void;
+  toSatisfy<E = any>(predicate: (x: E) => boolean): R;
 
   /**
    * Use `.toBeArray` when checking if a value is an `Array`.
    */
-  toBeArray(): void;
+  toBeArray(): R;
 
   /**
    * Use `.toBeArrayOfSize` when checking if a value is an `Array` of size x.
    * @param {Number} x
    */
-  toBeArrayOfSize(x: number): void;
+  toBeArrayOfSize(x: number): R;
 
   /**
    * Use `.toBeAfter` when checking if a date occurs after `date`.
    * @param {Date} date
    */
-  toBeAfter(date: Date): void;
+  toBeAfter(date: Date): R;
 
   /**
    * Use `.toBeBefore` when checking if a date occurs before `date`.
    * @param {Date} date
    */
-  toBeBefore(date: Date): void;
+  toBeBefore(date: Date): R;
 
   /**
    * Use `.toIncludeAllMembers` when checking if an `Array` contains all of the same members of a given set.
    * @param {Array.<*>} members
    */
-  toIncludeAllMembers<E = unknown>(members: readonly E[]): void;
+  toIncludeAllMembers<E = unknown>(members: readonly E[]): R;
 
   /**
    * Use `.toIncludeAnyMembers` when checking if an `Array` contains any of the members of a given set.
    * @param {Array.<*>} members
    */
-  toIncludeAnyMembers<E = unknown>(members: readonly E[]): void;
+  toIncludeAnyMembers<E = unknown>(members: readonly E[]): R;
 
   /**
    * Use `.toIncludeSameMembers` when checking if two arrays contain equal values, in any order.
    * @param {Array.<*>} members
    */
-  toIncludeSameMembers<E = unknown>(members: readonly E[]): void;
+  toIncludeSameMembers<E = unknown>(members: readonly E[]): R;
 
   /**
    * Use `.toPartiallyContain` when checking if any array value matches the partial member.
    * @param {*} member
    */
-  toPartiallyContain<E = unknown>(member: E): void;
+  toPartiallyContain<E = unknown>(member: E): R;
 
   /**
    * Use `.toSatisfyAll` when you want to use a custom matcher by supplying a predicate function that returns a `Boolean` for all values in an array.
    * @param {Function} predicate
    */
-  toSatisfyAll<E = any>(predicate: (x: E) => boolean): void;
+  toSatisfyAll<E = any>(predicate: (x: E) => boolean): R;
 
   /**
    * Use `.toSatisfyAny` when you want to use a custom matcher by supplying a predicate function that returns `true` for any matching value in an array.
    * @param {Function} predicate
    */
-  toSatisfyAny(predicate: (x: any) => boolean): void;
+  toSatisfyAny(predicate: (x: any) => boolean): R;
 
   /**
    * Use `.toBeBoolean` when checking if a value is a `Boolean`.
    */
-  toBeBoolean(): void;
+  toBeBoolean(): R;
 
   /**
    * Use `.toBeTrue` when checking a value is equal (===) to `true`.
    */
-  toBeTrue(): void;
+  toBeTrue(): R;
 
   /**
    * Use `.toBeFalse` when checking a value is equal (===) to `false`.
    */
-  toBeFalse(): void;
+  toBeFalse(): R;
 
   /**
    * Use `.toBeDate` when checking if a value is a `Date`.
    */
-  toBeDate(): void;
+  toBeDate(): R;
 
   /**
    * Use `.toBeValidDate` when checking if a value is a `valid Date`.
    */
-  toBeValidDate(): void;
+  toBeValidDate(): R;
 
   /**
    * Use `.toBeFunction` when checking if a value is a `Function`.
    */
-  toBeFunction(): void;
+  toBeFunction(): R;
 
   /**
    * Use `.toBeDateString` when checking if a value is a valid date string.
    */
-  toBeDateString(): void;
+  toBeDateString(): R;
 
   /**
    * Use `.toBeHexadecimal` when checking if a value is a valid HTML hex color.
    */
-  toBeHexadecimal(): void;
+  toBeHexadecimal(): R;
 
   /**
    * Use `.toHaveBeenCalledBefore` when checking if a `Mock` was called before another `Mock`.
@@ -146,7 +146,7 @@ interface CustomMatchers {
    * @param {Mock} mock
    * @param {boolean} [failIfNoSecondInvocation=true]
    */
-  toHaveBeenCalledBefore(mock: jest.MockInstance<unknown, unknown[]>, failIfNoSecondInvocation: boolean): void;
+  toHaveBeenCalledBefore(mock: jest.MockInstance<unknown, unknown[]>, failIfNoSecondInvocation: boolean): R;
 
   /**
    * Use `.toHaveBeenCalledAfter` when checking if a `Mock` was called after another `Mock`.
@@ -156,52 +156,52 @@ interface CustomMatchers {
    * @param {Mock} mock
    * @param {boolean} [failIfNoFirstInvocation=true]
    */
-  toHaveBeenCalledAfter(mock: jest.MockInstance<unknown, unknown[]>, failIfNoFirstInvocation: boolean): void;
+  toHaveBeenCalledAfter(mock: jest.MockInstance<unknown, unknown[]>, failIfNoFirstInvocation: boolean): R;
 
   /**
    * Use `.toHaveBeenCalledOnce` to check if a `Mock` was called exactly one time.
    */
-  toHaveBeenCalledOnce(): void;
+  toHaveBeenCalledOnce(): R;
 
   /**
    * Use `.toHaveBeenCalledOnceWith` to check if a `Mock` was called exactly one time with the expected value.
    */
-  toHaveBeenCalledOnceWith(): void;
+  toHaveBeenCalledOnceWith(): R;
 
   /**
    * Use `.toBeNumber` when checking if a value is a `Number`.
    */
-  toBeNumber(): void;
+  toBeNumber(): R;
 
   /**
    * Use `.toBeNaN` when checking a value is `NaN`.
    */
-  toBeNaN(): void;
+  toBeNaN(): R;
 
   /**
    * Use `.toBeFinite` when checking if a value is a `Number`, not `NaN` or `Infinity`.
    */
-  toBeFinite(): void;
+  toBeFinite(): R;
 
   /**
    * Use `.toBePositive` when checking if a value is a positive `Number`.
    */
-  toBePositive(): void;
+  toBePositive(): R;
 
   /**
    * Use `.toBeNegative` when checking if a value is a negative `Number`.
    */
-  toBeNegative(): void;
+  toBeNegative(): R;
 
   /**
    * Use `.toBeEven` when checking if a value is an even `Number`.
    */
-  toBeEven(): void;
+  toBeEven(): R;
 
   /**
    * Use `.toBeOdd` when checking if a value is an odd `Number`.
    */
-  toBeOdd(): void;
+  toBeOdd(): R;
 
   /**
    * Use `.toBeWithin` when checking if a number is in between the given bounds of: start (inclusive) and end (exclusive).
@@ -209,7 +209,7 @@ interface CustomMatchers {
    * @param {Number} start
    * @param {Number} end
    */
-  toBeWithin(start: number, end: number): void;
+  toBeWithin(start: number, end: number): R;
 
   /**
    * Use `.toBeInRange` when checking if an array has elements in range min (inclusive) and max (inclusive).
@@ -217,154 +217,154 @@ interface CustomMatchers {
    * @param min
    * @param max
    */
-  toBeInRange(min: number, max: number): void;
+  toBeInRange(min: number, max: number): R;
 
   /**
    * Use `.toBeObject` when checking if a value is an `Object`.
    */
-  toBeObject(): void;
+  toBeObject(): R;
 
   /**
    * Use `.toContainKey` when checking if an object contains the provided key.
    *
    * @param {String} key
    */
-  toContainKey(key: string): void;
+  toContainKey(key: string): R;
 
   /**
    * Use `.toContainKeys` when checking if an object has all of the provided keys.
    *
    * @param {Array.<String>} keys
    */
-  toContainKeys<E = unknown>(keys: readonly (keyof E | string)[]): void;
+  toContainKeys<E = unknown>(keys: readonly (keyof E | string)[]): R;
 
   /**
    * Use `.toContainAllKeys` when checking if an object only contains all of the provided keys.
    *
    * @param {Array.<String>} keys
    */
-  toContainAllKeys<E = unknown>(keys: readonly (keyof E | string)[]): void;
+  toContainAllKeys<E = unknown>(keys: readonly (keyof E | string)[]): R;
 
   /**
    * Use `.toContainAnyKeys` when checking if an object contains at least one of the provided keys.
    *
    * @param {Array.<String>} keys
    */
-  toContainAnyKeys<E = unknown>(keys: readonly (keyof E | string)[]): void;
+  toContainAnyKeys<E = unknown>(keys: readonly (keyof E | string)[]): R;
 
   /**
    * Use `.toContainValue` when checking if an object contains the provided value.
    *
    * @param {*} value
    */
-  toContainValue<E = unknown>(value: E): void;
+  toContainValue<E = unknown>(value: E): R;
 
   /**
    * Use `.toContainValues` when checking if an object contains all of the provided values.
    *
    * @param {Array.<*>} values
    */
-  toContainValues<E = unknown>(values: readonly E[]): void;
+  toContainValues<E = unknown>(values: readonly E[]): R;
 
   /**
    * Use `.toContainAllValues` when checking if an object only contains all of the provided values.
    *
    * @param {Array.<*>} values
    */
-  toContainAllValues<E = unknown>(values: readonly E[]): void;
+  toContainAllValues<E = unknown>(values: readonly E[]): R;
 
   /**
    * Use `.toContainAnyValues` when checking if an object contains at least one of the provided values.
    *
    * @param {Array.<*>} values
    */
-  toContainAnyValues<E = unknown>(values: readonly E[]): void;
+  toContainAnyValues<E = unknown>(values: readonly E[]): R;
 
   /**
    * Use `.toContainEntry` when checking if an object contains the provided entry.
    *
    * @param {Array.<[keyof E, E[keyof E]>} entry
    */
-  toContainEntry<E = unknown>(entry: readonly [keyof E, E[keyof E]]): void;
+  toContainEntry<E = unknown>(entry: readonly [keyof E, E[keyof E]]): R;
 
   /**
    * Use `.toContainEntries` when checking if an object contains all of the provided entries.
    *
    * @param {Array.<Array.<keyof E, E[keyof E]>>} entries
    */
-  toContainEntries<E = unknown>(entries: readonly (readonly [keyof E, E[keyof E]])[]): void;
+  toContainEntries<E = unknown>(entries: readonly (readonly [keyof E, E[keyof E]])[]): R;
 
   /**
    * Use `.toContainAllEntries` when checking if an object only contains all of the provided entries.
    *
    * @param {Array.<Array.<keyof E, E[keyof E]>>} entries
    */
-  toContainAllEntries<E = unknown>(entries: readonly (readonly [keyof E, E[keyof E]])[]): void;
+  toContainAllEntries<E = unknown>(entries: readonly (readonly [keyof E, E[keyof E]])[]): R;
 
   /**
    * Use `.toContainAnyEntries` when checking if an object contains at least one of the provided entries.
    *
    * @param {Array.<Array.<keyof E, E[keyof E]>>} entries
    */
-  toContainAnyEntries<E = unknown>(entries: readonly (readonly [keyof E, E[keyof E]])[]): void;
+  toContainAnyEntries<E = unknown>(entries: readonly (readonly [keyof E, E[keyof E]])[]): R;
 
   /**
    * Use `.toBeExtensible` when checking if an object is extensible.
    */
-  toBeExtensible(): void;
+  toBeExtensible(): R;
 
   /**
    * Use `.toBeFrozen` when checking if an object is frozen.
    */
-  toBeFrozen(): void;
+  toBeFrozen(): R;
 
   /**
    * Use `.toBeSealed` when checking if an object is sealed.
    */
-  toBeSealed(): void;
+  toBeSealed(): R;
 
   /**
    * Use `.toResolve` when checking if a promise resolves.
    */
-  toResolve(): void;
+  toResolve(): R;
 
   /**
    * Use `.toReject` when checking if a promise rejects.
    */
-  toReject(): void;
+  toReject(): R;
 
   /**
    * Use `.toBeString` when checking if a value is a `String`.
    */
-  toBeString(): void;
+  toBeString(): R;
 
   /**
    * Use `.toEqualCaseInsensitive` when checking if a string is equal (===) to another ignoring the casing of both strings.
    *
    * @param {String} string
    */
-  toEqualCaseInsensitive(string: string): void;
+  toEqualCaseInsensitive(string: string): R;
 
   /**
    * Use `.toStartWith` when checking if a `String` starts with a given `String` prefix.
    *
    * @param {String} prefix
    */
-  toStartWith(prefix: string): void;
+  toStartWith(prefix: string): R;
 
   /**
    * Use `.toEndWith` when checking if a `String` ends with a given `String` suffix.
    *
    * @param {String} suffix
    */
-  toEndWith(suffix: string): void;
+  toEndWith(suffix: string): R;
 
   /**
    * Use `.toInclude` when checking if a `String` includes the given `String` substring.
    *
    * @param {String} substring
    */
-  toInclude(substring: string): void;
+  toInclude(substring: string): R;
 
   /**
    * Use `.toIncludeRepeated` when checking if a `String` includes the given `String` substring the correct number of times.
@@ -372,14 +372,14 @@ interface CustomMatchers {
    * @param {String} substring
    * @param {Number} times
    */
-  toIncludeRepeated(substring: string, times: number): void;
+  toIncludeRepeated(substring: string, times: number): R;
 
   /**
    * Use `.toIncludeMultiple` when checking if a `String` includes all of the given substrings.
    *
    * @param {Array.<String>} substring
    */
-  toIncludeMultiple(substring: readonly string[]): void;
+  toIncludeMultiple(substring: readonly string[]): R;
 
   /**
    * Use `.toThrowWithMessage` when checking if a callback function throws an error of a given type with a given error message.
@@ -387,43 +387,43 @@ interface CustomMatchers {
    * @param {Function} type
    * @param {String | RegExp} message
    */
-  toThrowWithMessage(type: (...args: any[]) => any, message: string | RegExp): void;
+  toThrowWithMessage(type: (...args: any[]) => any, message: string | RegExp): R;
 
   /**
    * Use `.toBeEmptyObject` when checking if a value is an empty `Object`.
    */
-  toBeEmptyObject(): void;
+  toBeEmptyObject(): R;
 
   /**
    * Use `.toBeSymbol` when checking if a value is a `Symbol`.
    */
-  toBeSymbol(): void;
+  toBeSymbol(): R;
 
   /**
    * Use `.toBeBetween` when checking if a date occurs between `startDate` and `endDate`.
    * @param {Date} startDate
    * @param {Date} endDate
    */
-  toBeBetween(startDate: Date, endDate: Date): void;
+  toBeBetween(startDate: Date, endDate: Date): R;
 
   /**
    * Use `.toBeBeforeOrEqualTo` when checking if a date equals to or occurs before `date`.
    * @param {Date} date
    */
-  toBeBeforeOrEqualTo(date: Date): void;
+  toBeBeforeOrEqualTo(date: Date): R;
 
   /**
    * Use `.toBeAfterOrEqualTo` when checking if a date equals to or occurs after `date`.
    * @param {Date} date
    */
-  toBeAfterOrEqualTo(date: Date): void;
+  toBeAfterOrEqualTo(date: Date): R;
 
   /**
    * Use `.toEqualIgnoringWhitespace` when checking if a `String` is equal (===) to given `String` ignoring white-space.
    *
    * @param {String} string
    */
-  toEqualIgnoringWhitespace(string: string): void;
+  toEqualIgnoringWhitespace(string: string): R;
 }
 
 declare namespace jest {
@@ -873,7 +873,7 @@ declare namespace jest {
 
   // noinspection JSUnusedGlobalSymbols
   // eslint-disable-next-line @typescript-eslint/no-empty-interface
-  interface Expect extends CustomMatchers {}
+  interface Expect extends CustomMatchers<AsymmetricMatcher> {}
 
   // noinspection JSUnusedGlobalSymbols
   // eslint-disable-next-line @typescript-eslint/no-empty-interface
@@ -882,5 +882,5 @@ declare namespace jest {
 
 declare namespace Vi {
   // eslint-disable-next-line @typescript-eslint/no-empty-interface
-  interface AsymmetricMatchersContaining extends CustomMatchers {}
+  interface AsymmetricMatchersContaining extends CustomMatchers<any> {}
 }

--- a/types/index.d.ts
+++ b/types/index.d.ts
@@ -873,7 +873,7 @@ declare namespace jest {
 
   // noinspection JSUnusedGlobalSymbols
   // eslint-disable-next-line @typescript-eslint/no-empty-interface
-  interface Expect extends CustomMatchers<AsymmetricMatcher> {}
+  interface Expect extends CustomMatchers<any> {}
 
   // noinspection JSUnusedGlobalSymbols
   // eslint-disable-next-line @typescript-eslint/no-empty-interface

